core/async_cell.rs

```diff
@@ -0,0 +1,619 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+use std::any::Any;
+use std::borrow::Borrow;
+use std::cell::Cell;
+use std::cell::UnsafeCell;
+use std::collections::VecDeque;
+use std::ops::Deref;
+use std::rc::Rc;
+
+use self::internal as i;
+
+pub type AsyncRef<T> = i::AsyncBorrowImpl<T, i::Shared>;
+pub type AsyncMut<T> = i::AsyncBorrowImpl<T, i::Exclusive>;
+
+pub type AsyncRefFuture<T> = i::AsyncBorrowFutureImpl<T, i::Shared>;
+pub type AsyncMutFuture<T> = i::AsyncBorrowFutureImpl<T, i::Exclusive>;
+
+pub struct AsyncRefCell<T> {
+  value: UnsafeCell<T>,
+  borrows: Cell<i::BorrowCount>,
+  waiters: Cell<VecDeque<Option<i::Waiter>>>,
+  turn: Cell<usize>,
+}
+
+impl<T: 'static> AsyncRefCell<T> {
+  /// Create a new `AsyncRefCell` that encapsulates the specified value.
+  /// Note that in order to borrow the inner value, the `AsyncRefCell`
+  /// needs to be wrapped in an `Rc` or an `RcRef`. These can be created
+  /// either manually, or by using the convenience method
+  /// `AsyncRefCell::new_rc()`.
+  pub fn new(value: T) -> Self {
```

Comment:
If it is required to wrap in an `Rc` do we need this method? Why not make `new_rc` the default?

Reply:
Look at http_bench_bin_ops: a `TcpStream` contains two `AsyncRefCell`s but they're not individually wrapped in an `Rc` - just the struct that contains them is. So it needs to be possible to create a "naked" `AsyncRefCell`.